### PR TITLE
[Merged by Bors] - feat(Mathlib/LinearAlgebra/DirectSum/Finsupp): tensor products of finsupp sums

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
## Modules

* `TensorProduct.finsuppLeft`, the tensor product of `ι →₀ M` and `N` is linearly equivalent to `ι →₀ M ⊗[R] N`

* `TensorProduct.finsuppScalarLeft`, the tensor product of `ι →₀ R` and `N` is linearly equivalent to `ι →₀ N`

* `TensorProduct.finsuppRight`, the tensor product of `M` and `ι →₀ N` is linearly equivalent to `ι →₀ M ⊗[R] N`

* `TensorProduct.finsuppLeft'`, if `M` is an `S`-module, then the tensor product of `ι →₀ M` and `N` is `S`-linearly equivalent to `ι →₀ M ⊗[R] N`

This is the first part of PR #10824 which contains three applications of these functions to monoid algebras, polynomials and multivariate polynomials.

It has been split because this part is reasonably sound, while the three other files are more like propositions.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by acmepjz** [comment]
Thank you very much! It looks good for me.

**Comment by AntoineChambert-Loir** [comment]
!bench

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean
Comment (alreadydone) [inline Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:155]: I'd remove unnecessary parentheses (you did not add parentheses in the previous lemma for example).

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean
Comment (alreadydone) [inline Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:86]: I'd replace `.trans` by `≪≫ₗ`. Do you prefer `trans`?

**Review by alreadydone — COMMENTED** [review]
Largely LGTM but I might come back to golf some proofs ...

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:86]: yes, it's much faster to type to me…

Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:86]: Anyway, I changed trans to that. It spares some parentheses…

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean
Comment (AntoineChambert-Loir) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean]: OK

**Review by alreadydone — COMMENTED** [review]
Sorry for the delay, I'll get back to this tomorrow.

**Review by alreadydone — APPROVED** [review]
I've done the golfs [here](https://github.com/leanprover-community/mathlib4/compare/ACL/FinsuppTensorProd_part1...jyxu/FinsuppTensorProd?expand=1). I made the defs `finsuppLeft` etc. shorter but it may make it harder for the reader, so feel free to adopt the changes that you think is appropriate. I gave two proofs for `finsuppLeft_apply_tmul` etc. and you're free to adopt either.

It's a bit weird that `finsuppLeft` etc. have no explicit argument, but LGTM.

**Comment by AntoineChambert-Loir** [comment]
Thanks for the shorter versions, I'll use them, and I'm OK to make the arguments of finsuppLeft explicit, in accordance with what is done for docs#TensorProduct.directSumLeft

**Comment by alreadydone** [comment]
Can you please fix the compile errors (which probably results from the implicitness changes)? Thanks!

**Comment by AntoineChambert-Loir** [comment]
I had not noticed that there were errors left, sorry. I make the necessary adjustments at once.

**Comment by alreadydone** [comment]
Thanks 🎉 
maintainer merge

**Comment by ocfnash** [comment]
Thanks!

bors merge

**Pull request closed without merging**
